### PR TITLE
Implement the user nickname feature that allows the user to fill in their nickname, address and phone number after loggin into their account

### PR DESCRIPTION
**Pull request by LeonardZ0731**

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/UserInfoServlet.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    if (nickname.equals("") || address.equals("") || phone.equals("")) {
+      System.err.println("At least one input field is empty");
+      response.sendRedirect("/400.html");
+      return;
+    }
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query("UserInfo").setFilter(new FilterPredicate("email", FilterOperator.EQUAL, email));
```
Comment (poncem91) [inline src/main/java/com/google/sps/servlets/UserInfoServlet.java:50]: Same comment as before about using the userId to query users instead :)

File: src/main/java/com/google/sps/servlets/UserInfoServlet.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      entity = new Entity("UserInfo");
```
Comment (poncem91) [inline src/main/java/com/google/sps/servlets/UserInfoServlet.java:54]: It'd be helpful to store the userId in here too

Comment (LeonardZ0731) [reply src/main/java/com/google/sps/servlets/UserInfoServlet.java:54]: Sure. I would modify it to userId in the next commit.

File: src/main/webapp/index.jsp [hunk 3]
```diff
@@ -49,7 +49,7 @@
           	      <p class="login-messages"><%=userService.getCurrentUser().getEmail()%> | <a href="<%=logoutUrl%>">Logout</a></p>
                   <%
                   } else {
-                      String urlToRedirectToAfterUserLogsIn = "/";
+                      String urlToRedirectToAfterUserLogsIn = "/account.jsp";
```
[the remark this answers is not in the document]
Comment (LeonardZ0731) [reply src/main/webapp/index.jsp:52]: In `account.jsp`, I checked whether the user already has details saved. So even if the user get redirected to `account.jsp`, he will be redirected to `user_profile.jsp` at the end.

File: src/main/webapp/account.jsp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        <div style="clear: both"><div/>
+        <hr/>
+        <div id="form-container">
+            <form id="new-user-info-form" action="/account" method="POST">
+                <div>
+                    <label for="nickname-input">Your Preferred Nickname:</label>
+                    <br/>
+                </div>
+                <br/>
+                <textarea name="nickname-input" id="nickname-input" placeholder="Input your preferred nickname here:"></textarea>
```
[the remark this answers is not in the document]
Comment (LeonardZ0731) [reply src/main/webapp/account.jsp:29]: The `required="true"` attribute cannot prevent users from submitting empty spaces. Thus, I decided to first take in the results, then use the `string.trim()` method to remove the trailing spaces then check whether it is equal to empty string. I think this method can be better, but adding a `required` attribute can also be great. I will add them in the new commit.

[the remark this answers is not in the document]
Comment (poncem91) [reply src/main/webapp/account.jsp:29]: Ah gotcha! I've never tried this myself before, but I think you can also use regex to validate input patterns within the html form itself.  But between the required="true" attribute and the method you had already implemented we're probably good anyway!

File: src/main/webapp/account.jsp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  <% UserService userService = UserServiceFactory.getUserService();
+  if (!userService.isUserLoggedIn()) {
+       response.sendRedirect(userService.createLoginURL("/account.jsp"));
+  } else if (RetrieveInfo.getInfo(userService) == null) { %>
+  <body>
+    <div id="container">
+        <div id="header">
+            <h1 id="title">Please input your personal information</h1>
+        </div>
+        <div style="clear: both"><div/>
```
[the remark this answers is not in the document]
Comment (LeonardZ0731) [reply src/main/webapp/account.jsp:20]: Sure. I will put them in the css file in the new commit.

File: src/main/java/com/google/sps/helper/RetrieveInfo.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.users.UserService;
+import java.util.ArrayList;
+import java.util.List;
+
+public final class RetrieveInfo {
```
[the remark this answers is not in the document]
Comment (LeonardZ0731) [reply src/main/java/com/google/sps/helper/RetrieveInfo.java:28]: That's a better name for the class! I will modify them in the new commit.

File: src/main/java/com/google/sps/servlets/UserInfoServlet.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+public class UserInfoServlet extends HttpServlet {
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect(userService.createLoginURL("/account.jsp"));
+      return;
+    }
+
+    String nickname = "";
```
[the remark this answers is not in the document]
Comment (LeonardZ0731) [reply src/main/java/com/google/sps/servlets/UserInfoServlet.java:30]: Maybe one variable does work but from my perspective, one variable is a potential input and the other is the final result. So I would prefer to separate the potential input and final result to two variables.

File: src/main/java/com/google/sps/servlets/UserInfoServlet.java (hunk 1 above)
Comment (LeonardZ0731) [reply src/main/java/com/google/sps/servlets/UserInfoServlet.java:50]: Sure. I will change it in the new commit.

File: src/main/java/com/google/sps/servlets/TaskServlet.java [hunk 8]
```diff
@@ -68,7 +77,7 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     taskEntity.setProperty("timestamp", creationTime);
     taskEntity.setProperty("reward", rewardPts);
     taskEntity.setProperty("status", "OPEN");
-    taskEntity.setProperty("Owner", "Leonard");
+    taskEntity.setProperty("Owner", userService.getCurrentUser().getEmail());
```
[the remark this answers is not in the document]
Comment (LeonardZ0731) [reply src/main/java/com/google/sps/servlets/TaskServlet.java:80]: I will change the email to `userId` instead in the new commit. Based on the previous comment, I think using the `userId` here is a better choice.

File: src/main/java/com/google/sps/helper/RetrieveInfo.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  public static List<String> getInfoFromEmail(String email) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query("UserInfo").setFilter(new FilterPredicate("email", FilterOperator.EQUAL, email));
```
[the remark this answers is not in the document]
Comment (LeonardZ0731) [reply src/main/java/com/google/sps/helper/RetrieveInfo.java:32]: Oh really? That's surprising 🤣 . I will change it to userId in the new commit.

**Pull request merged**